### ISSUE TITLE
BUG: pd.sort_values() not working on categorical column

### ISSUE THREAD

**Issue by BelaNeuendorf** [pandas-dev/pandas#58153]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame([[1, 2, 'March'],[5, 6, 'Dec'],[3, 4, 'April']], columns=['a','b','month'])
df.month = df.month.astype('category')

custom_dict = {'March': 0, 'April': 1, 'Dec': 3}

df.sort_values(by=['month'], key=lambda x: x.map(custom_dict))
```

### Issue Description

the sorting is done alphabetically instead of by the provided dictionary.
Result:
	a	b	month
2	3	4	April
1	5	6	Dec
0	1	2	March

### Expected Behavior

<html>
<body>

  | a | b | m
-- | -- | -- | --
1 | 2 | March
3 | 4 | April
5 | 6 | Dec

</body>
</html>

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.12.2.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 11
Version               : 10.0.22631
machine               : AMD64
processor             : AMD64 Family 25 Model 80 Stepping 0, AuthenticAMD
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : de_DE.cp1252

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2023.3.post1
dateutil              : 2.8.2
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : None
pytest                : None
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : None
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : 8.22.2
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : None
bottleneck            : 1.3.7
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.8.0
numba                 : None
numexpr               : 2.8.7
odfpy                 : None
openpyxl              : None
pandas_gbq            : None
pyarrow               : None
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.12.0
sqlalchemy            : None
tables                : None
tabulate              : None
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None

</details>

**Comment by duchuyvp** [comment]
I'll look at it if it was a bug. Currently you can use this as an alternative
```python
df.sort_values(by="month", key=lambda x: x.map(custom_dict).astype(int))
```

**Comment by ahmedembeddedx** [comment]
import pandas as pd

# Create DataFrame
df = pd.DataFrame([[1, 2, 'March'],[5, 6, 'Dec'],[3, 4, 'April']], columns=['a','b','month'])

# Define custom_dict
custom_dict = {'March': 0, 'April': 1, 'Dec': 3}

# Convert 'month' column to Categorical with custom order
df['month'] = pd.Categorical(df['month'], categories=custom_dict.keys())

# Sort DataFrame based on the custom order
df.sort_values(by='month', key=lambda x: x.map(custom_dict))

**Comment by rtlee9** [comment]
To be more specific, `sort_values` doesn't respect the `key` parameter for categorical series.

Sorting does work as expected if the categorical is ordered using the `ordered` param:

```python
import pandas as pd

df = pd.DataFrame([[1, 2, 'March'],[5, 6, 'Dec'],[3, 4, 'April']], columns=['a','b','month'])
df['month_cat_ordered'] = pd.Categorical(df.month, ['March', 'April', 'Dec'], ordered=True)
df.sort_values('month_cat_ordered')
```
|    |   a |   b | month   | month_cat_ordered   |                                                                                                           
|---:|----:|----:|:--------|:--------------------|                                                                                                  
|  0 |   1 |   2 | March   | March               |                                                                                        
|  2 |   3 |   4 | April   | April               |                                                                              
|  1 |   5 |   6 | Dec     | Dec                 |       

Mapping an unordered categorical as in the original example maps the old categories to the new categories but doesn't reorder the new categories. This is possible to implement this reordering in `Categorical.map` but would come at a performance cost and it isn't immediately clear it's worth it.

**Comment by jorisvandenbossche** [comment]
> Mapping an unordered categorical as in the original example maps the old categories to the new categories but doesn't reorder the new categories.

Indeed, the mapped values (in this case the integers to sort by) are re-interpreted again as a categorical with the same order as the original categories. That gives a clearly unintended behaviour.
